File: src/main/java/cloud/cosmin/checklister/rest/ItemController.java
package cloud.cosmin.checklister.rest;

import cloud.cosmin.checklister.dao.ItemEntity;
import cloud.cosmin.checklister.dto.ItemGetDto;
import cloud.cosmin.checklister.dto.ItemPostDto;
import cloud.cosmin.checklister.repo.ItemRepo;
import cloud.cosmin.checklister.service.ConverterService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;

import java.util.Optional;
import java.util.UUID;

@Controller
public class ItemController {
    @Autowired
    private ItemRepo itemRepo;

    @Autowired
    private ConverterService converterService;

    @GetMapping("/api/v1/item/{itemId}")
    public ResponseEntity<ItemGetDto> getListItem(@PathVariable UUID itemId) {
        if(itemId == null) {
            return ResponseEntity.badRequest().build();
        }

        Optional<ItemEntity> optionalItem = itemRepo.findById(itemId);

        if(!optionalItem.isPresent()) {
            return ResponseEntity.notFound().build();
        }

        ItemGetDto dto = converterService.itemDto(optionalItem.get());
        return ResponseEntity.ok(dto);
    }

    @PutMapping("/api/v1/item/{itemId}")
    public ResponseEntity<ItemGetDto> updateItem(@PathVariable UUID itemId,
                                                 @RequestBody ItemPostDto itemPost) {
        if(itemId == null) {
            return ResponseEntity.badRequest().build();
        }

        Optional<ItemEntity> optionalItem = itemRepo.findById(itemId);
        if(!optionalItem.isPresent()) {
            return ResponseEntity.notFound().build();
        }

        ItemEntity item = optionalItem.get();
        item.setContent(itemPost.content);
        item.setContentType(itemPost.contentType);

        ItemEntity saved = itemRepo.save(item);
        ItemGetDto dto = converterService.itemDto(saved);
        return ResponseEntity.ok(dto);

    }
}

File: src/main/kotlin/cloud/cosmin/checklister/rest/ItemController.kt
package cloud.cosmin.checklister.rest

import cloud.cosmin.checklister.lib.dto.ItemGetDto
import cloud.cosmin.checklister.lib.dto.ItemPostDto
import cloud.cosmin.checklister.lib.event.model.RankOperation
import cloud.cosmin.checklister.service.ItemService
import io.swagger.annotations.Api
import io.swagger.annotations.ApiOperation
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.util.*

@Controller
@Api(description = "Operations on list items", tags = arrayOf("item"))
class ItemController
@Autowired
constructor(private val itemService: ItemService) {

    @GetMapping("/api/v1/item/{itemId}")
    @ApiOperation("Retrieve an item")
    fun getListItem(@PathVariable itemId: UUID?): ResponseEntity<ItemGetDto> {
        if (itemId == null) {
            return ResponseEntity.badRequest().build()
        }

        val optionalItem = itemService.findById(itemId)

        if (!optionalItem.isPresent) {
            return ResponseEntity.notFound().build()
        }

        return ResponseEntity.ok(optionalItem.get())
    }

    @PostMapping("/api/v1/item")
    @ApiOperation("Create an item")
    fun createItem(@RequestBody dto: ItemPostDto): ResponseEntity<ItemGetDto> {
        val createdItem = itemService.create(dto)
        return ResponseEntity
                .created(URI.create("/api/v1/item/" + createdItem.id))
                .body(createdItem)
    }


    @PutMapping("/api/v1/item/{itemId}")
    @ApiOperation("Update an item")
    fun updateItem(@PathVariable itemId: UUID,
                   @RequestBody dto: ItemPostDto): ResponseEntity<ItemGetDto> {
        val updated = itemService.update(itemId, dto)
        return ResponseEntity.ok(updated)
    }

    @PostMapping("/api/v1/item/{itemId}/rank/up")
    @ApiOperation("Move the item one rank up in its list")
    fun rankUp(@PathVariable itemId: UUID): ResponseEntity<ItemGetDto> {
        return ResponseEntity.ok(itemService.rank(itemId, RankOperation.UP))
    }

    @PostMapping("/api/v1/item/{itemId}/rank/down")
    @ApiOperation("Move the item one rank down in its list")
    fun rankDown(@PathVariable itemId: UUID): ResponseEntity<ItemGetDto> {
        return ResponseEntity.ok(itemService.rank(itemId, RankOperation.DOWN))
    }

    @PostMapping("/api/v1/item/{itemId}/rank/top")
    @ApiOperation("Move the item to the top of its list")
    fun rankTop(@PathVariable itemId: UUID): ResponseEntity<ItemGetDto> {
        return ResponseEntity.ok(itemService.rank(itemId, RankOperation.TOP))
    }

    @PostMapping("/api/v1/item/{itemId}/rank/bottom")
    @ApiOperation("Move the item to the bottom of its list")
    fun rankBottom(@PathVariable itemId: UUID): ResponseEntity<ItemGetDto> {
        return ResponseEntity.ok(itemService.rank(itemId, RankOperation.BOTTOM))
    }
}
